lightning/src/events/mod.rs

```diff
@@ -270,6 +270,41 @@ impl_writeable_tlv_based_enum!(InterceptNextHop,
 	};
 );
 
+/// The reason the payment failed. Used in [`Event::PaymentFailed`].
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum PaymentFailureReason {
+	/// The intended recipient rejected our payment.
+	RecipientRejected,
+	/// The payment failed while being forwarded along the path. A [`Event::PaymentPathFailed`]
+	/// should have come before this.
+	FailedAlongPath,
```

Comment:
I wonder if we should consolidate this with `ExhaustedRetryAttempts`, because in the 1 place we return this error, I think retries being exhausted is arguably the "real" reason for the `PaymentFailed` event, and the `FailedAlongPath` part is communicated via the `PaymentPathFailed` events. cc @TheBlueMatt does that make sense?

Reply:
> Sadly doing this is going to require storing the reason in the `Abandoned` payment case - we don't know without context whether we were `UserAbandoned` or `ExhaustedRetryAttempts`.

Just to clarify did you mean `RecipientRejected` here instead of `UserAbandoned`?

As I was looking into trying to store a reason in `Abandoned` it seemed like we didn't really have what we needed to store the correct reason for what was happening, which has led me to a question. I'm sort of confused on what actually would differentiate a situation where a recipient rejected our payment versus our path failed and we exhausted our last retry. When a recipient rejects our payment, we also get a path failure don't we (just with different onion error data)?

Reply:
Hmm, I'm confused about how we can't tell between user abandoned/exhausted. I'm thinking that the only time that `UserAbandoned` is used is in `ChannelManager::abandon_payment` which just calls `OutboundPayments::abandon_payment` which directly pushes the `PaymentFailed` with that reason?

Reply:
Yep, sounds good